osquery/filesystem/linux/proc.cpp

```diff
@@ -16,54 +16,165 @@
 #include <osquery/filesystem.h>
 #include <osquery/logger.h>
 
+#include "osquery/core/conversions.h"
+#include "osquery/filesystem/linux/proc.h"
+
 namespace osquery {
+const char* kLinuxProcPath = "/proc";
 
-const std::string kLinuxProcPath = "/proc";
+Status procGetProcessNamespaces(const std::string& process_id,
+                                ProcessNamespaceList& namespace_list) {
+  namespace_list.clear();
 
-Status procProcesses(std::set<std::string>& processes) {
-  // Iterate over each process-like directory in proc.
-  boost::filesystem::directory_iterator it(kLinuxProcPath), end;
-  try {
-    for (; it != end; ++it) {
-      if (boost::filesystem::is_directory(it->status())) {
-        // See #792: std::regex is incomplete until GCC 4.9
-        if (std::atoll(it->path().leaf().string().c_str()) > 0) {
-          processes.insert(it->path().leaf().string());
-        }
-      }
+  for (std::string namespace_name :
+       {"cgroup", "ipc", "mnt", "net", "pid", "user", "uts"}) {
+    std::string symlink_path = std::string(kLinuxProcPath) + "/" + process_id +
+                               "/ns/" + namespace_name;
+
+    char link_destination[PATH_MAX] = {};
+    auto link_dest_length =
+        readlink(symlink_path.data(), link_destination, PATH_MAX - 1);
+    if (link_dest_length < 0) {
+      return Status(1,
+                    std::string("Failed to retrieve the inode for namespace ") +
+                        namespace_name + " in process " + process_id);
+    }
+
+    // The link destination must be in the following form: namespace:[inode]
+    if (std::strncmp(link_destination,
+                     namespace_name.data(),
+                     namespace_name.size()) != 0 ||
+        std::strncmp(link_destination + namespace_name.size(), ":[", 2) != 0) {
+      return Status(1,
+                    std::string("Invalid descriptor for namespace ") +
+                        namespace_name + " in process " + process_id);
     }
-  } catch (const boost::filesystem::filesystem_error& e) {
-    VLOG(1) << "Exception iterating Linux processes " << e.what();
-    return Status(1, e.what());
+
+    // Parse the inode part of the string; strtoull should return us a pointer
+    // to the
+    // closing square bracket
+    const char* inode_string_ptr = link_destination + namespace_name.size() + 2;
+    char* square_bracket_ptr = nullptr;
+
+    auto inode = static_cast<ino_t>(
+        std::strtoull(inode_string_ptr, &square_bracket_ptr, 10));
+    if (inode == 0 || square_bracket_ptr == nullptr ||
+        *square_bracket_ptr != ']') {
+      return Status(
+          1,
+          std::string("Invalid inode value in descriptor for namespace ") +
+              namespace_name + " in process " + process_id);
+    }
+
+    namespace_list[namespace_name] = inode;
   }
 
   return Status(0, "OK");
 }
 
+std::string procDecodeAddressFromHex(const std::string& encoded_address,
```

Comment:
Are we sure we haven't done this somewhere else in the code base?

Reply:
I think I just moved this function into this .cpp file (it was already defined somewhere else)

Reply:
This code was previously used in process_open_sockets.cpp; I have made a function out of it and moved it in proc.h/proc.cpp